Firebase.agent.lib.nut

```diff
@@ -641,24 +655,53 @@ class Firebase {
                 }
                 if (200 <= res.statuscode && res.statuscode < 300) {
                     onSuccess(data);
+                    _tooManyReqTimer = false;
+                } else if (res.statuscode == 28 || res.statuscode == 429) {
+                    local now = time();
+                    // Too many requests, set _tooManyReqTimer to prevent more requests to FB
+                    if (_tooManyReqTimer == false) {
+                        // This is the first 429 we have seen set a default timeout
+                        _tooManyReqTimer = now + DEFAULT_BACK_OFF_TIMEOUT_SEC;
+                    } else if (_tooManyReqTimer <= now) {
+                        // Firebase is still overwhelmed after first timeout expired, 
+                        // Let's block requests for longer to let FB recover
+                        _tooManyReqTimer = now + (DEFAULT_BACK_OFF_TIMEOUT_SEC * _tooManyReqCounter++);
+                    }
+                    // Pass error to callback
+                    onError("Error " + res.statuscode);
                 } else if (typeof data == "table" && "error" in data) {
+                    _tooManyReqTimer = false;
                     local error = data ? data.error : null;
                     onError(error);
                 } else {
+                    _tooManyReqTimer = false;
                     onError("Error " + res.statuscode);
                 }
             } catch (err) {
+                _tooManyReqTimer = false;
                 onError(err);
             }
         }
     }
 
-    // use Promise if promise libary included and callback != null
     function _processResponse(request, callback) {
-        if (_promiseIncluded && callback == null) {
-            return _createRequestPromise(request);
+        // Use Promise if promise libary included and callback != null
+        local usePromise = (_promiseIncluded && callback == null);
```

Comment:
I know it's not your code, but I'm wondering where is this check for callback comes from? Why can't the callback be executed from a promise? Do you have an idea?

Reply:
The library should function if the Promise library is not included. As I understand it if you include a callback that will be used, if you don't include a callback but have used the Promise library the CRUD methods will return a Promise.  I have not fully tested this.

Reply:
I see. Thanks.